cli/tests/upgrade.bats

```diff
@@ -166,3 +166,32 @@ EOF
   assert_success
   assert_output --partial "No packages need to be upgraded"
 }
+
+
+# Catalog functionality tests
+# ---------------------------------------------------------------------------- #
+
+# bats test_tags=upgrade:catalog
+@test "'flox install' installs package from catalog and builds it" {
+  export FLOX_FEATURES_USE_CATALOG=true
+
+  "$FLOX_BIN" init
+  # create a catalog manifest
+  echo "version = 1" > ".flox/env/manifest.toml"
+  echo 'options.systems = ["aarch64-darwin", "x86_64-darwin", "aarch64-linux", "x86_64-linux"]' >> ".flox/env/manifest.toml"
+
+  _FLOX_USE_CATALOG_MOCK="$TESTS_DIR/catalog_responses/hello_resolution_old.json" \
+  "$FLOX_BIN" install -i hello_install_id hello
+
+  run "$FLOX_BIN" list
+  assert_success
+  assert_line "hello_install_id: hello (old_version)"
+
+  _FLOX_USE_CATALOG_MOCK="$TESTS_DIR/catalog_responses/hello_resolution.json" \
+  run "$FLOX_BIN" upgrade hello_install_id
+  assert_success
+
+  run "$FLOX_BIN" list
+  assert_success
+  assert_line "hello_install_id: hello (version)"
```

Comment:
I think this is equivalent to tests of `Upgrade::handle`, but unlike for `*::handle` we actually have some setup "fixtures" in place for bats tests.

Unlike `upgrade_with_catalog_client` its actually testing whether the right upgrade implementation is used though i admit we can have that be a unit test of CoreEnvironment::upgrade as well.

Reply:
I added tests that check whether a catalog client is correcrtly required when we use a v1 manifest.
Personally a bats smoke test like that is still useful since that's the only place that tests user side of the program.